manta-pay/src/crypto/hash/poseidon/parameter_generation/lfsr.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! LFSH implementation.
+// adapted from: https://github.com/arkworks-rs/sponge/blob/51d6fc9aac1fa69f44a04839202b5de828584ed8/src/poseidon/grain_lfsr.rs
+
+use alloc::vec::Vec;
+
+use crate::crypto::hash::ParamField;
+
+pub struct GrainLFSR {
+    state: [bool; 80],
+    prime_num_bits: u64,
+    head: usize,
+}
+
+fn append_bits<T: Into<u128>>(state: &mut [bool; 80], head: &mut usize, n: usize, from: T) {
+    let val = from.into() as u128;
+    for i in (0..n).rev() {
+        state[*head] = (val >> i) & 1 != 0;
+        *head += 1;
+        *head %= 80;
+    }
+}
+
+impl GrainLFSR {
+    pub fn new(prime_num_bits: u64, width: usize, r_f: usize, r_p: usize) -> Self {
+        let mut init_sequence = [false; 80];
+        let mut head = 0;
+        // b0, b1 describes the field
+        append_bits(&mut init_sequence, &mut head, 2, 1u8);
+        // b2...=b5 describes s-box: we always use non-inverse s-box
+        append_bits(&mut init_sequence, &mut head, 4, 0b00000u8);
+        // b6...=b17 describes prime_num_bits
+        append_bits(&mut init_sequence, &mut head, 12, prime_num_bits);
+        // b18...=b29 describes width
+        append_bits(&mut init_sequence, &mut head, 12, width as u16);
+        // b30..=39 describes r_f (num_full_rounds)
+        append_bits(&mut init_sequence, &mut head, 10, r_f as u16);
+        // b40..=49 describes r_p (num_partial_rounds)
+        append_bits(&mut init_sequence, &mut head, 10, r_p as u16);
+        // b50..=79 describes the constant 1
+        append_bits(
+            &mut init_sequence,
+            &mut head,
+            30,
+            0b111111111111111111111111111111u128,
+        );
+        let mut res = GrainLFSR {
+            state: init_sequence,
+            prime_num_bits,
+            head,
+        };
+        res.init();
+        res
+    }
+
+    fn update(&mut self) -> bool {
+        let new_bit =
+            self.bit(62) ^ self.bit(51) ^ self.bit(38) ^ self.bit(23) ^ self.bit(13) ^ self.bit(0);
+        self.state[self.head] = new_bit;
+        self.head += 1;
+        self.head %= 80;
+        new_bit
+    }
+
+    fn init(&mut self) {
+        for _ in 0..160 {
+            self.update();
+        }
+    }

```

Comment:
This function is only called in the constructor. It's ok to separate it to manage complexity, but since you have control over the interface, design it to reduce code in the constructor. For example:

```suggestion
    /// ADD DOCS HERE
    fn init(mut self) -> Self {
        for _ in 0..160 {
            self.update();
        }
        self
    }
```

Also, remove any magic constants like `160` and move them to `struct` `const`s.

Reply:
Makes sense. I suggest adding constant `LFSR_SIZE`, and 160 will become `LFSR_SIZE * 2`